lib/src/body.dart

```diff
@@ -18,31 +19,70 @@ class Body {
   /// This will be `null` after [read] is called.
   Stream<List<int>> _stream;
 
-  Body._(this._stream);
+  /// The encoding used to encode the stream returned by [read], or `null` if no
+  /// encoding was used.
+  final Encoding encoding;
+
+  /// The length of the stream returned by [read], or `null` if that can't be
+  /// determined efficiently.
+  final int contentLength;
+
+  Body._(this._stream, this.encoding, this.contentLength);
 
   /// Converts [body] to a byte stream and wraps it in a [Body].
   ///
-  /// [body] may be either a [Body], a [String], a [Stream<List<int>>], or
-  /// `null`. If it's a [String], [encoding] will be used to convert it to a
-  /// [Stream<List<int>>].
+  /// [body] may be either a [Body], a [String], a [List<int>], a
+  /// [Stream<List<int>>], or `null`. If it's a [String], [encoding] will be
+  /// used to convert it to a [Stream<List<int>>].
   factory Body(body, [Encoding encoding]) {
-    if (encoding == null) encoding = UTF8;
-
     if (body is Body) return body;
 
     Stream<List<int>> stream;
+    int contentLength;
     if (body == null) {
+      contentLength = 0;
       stream = new Stream.fromIterable([]);
     } else if (body is String) {
-      stream = new Stream.fromIterable([encoding.encode(body)]);
+      if (encoding == null) {
+        var encoded = UTF8.encode(body);
+        // If the text is plain ASCII, don't modify the encoding. This means
+        // that an encoding of "text/plain" will stay put.
+        if (!_isPlainAscii(encoded, body.length)) encoding = UTF8;
+        contentLength = encoded.length;
+        stream = new Stream.fromIterable([encoded]);
+      } else {
+        var encoded = encoding.encode(body);
+        contentLength = encoded.length;
+        stream = new Stream.fromIterable([encoded]);
+      }
+    } else if (body is List) {
+      contentLength = body.length;
+      stream = new Stream.fromIterable([DelegatingList.typed(body)]);
     } else if (body is Stream) {
       stream = DelegatingStream.typed(body);
     } else {
       throw new ArgumentError('Response body "$body" must be a String or a '
           'Stream.');
     }
 
-    return new Body._(stream);
+    return new Body._(stream, encoding, contentLength);
+  }
+
+  /// Returns whether [bytes] is plain ASCII.
+  ///
+  /// [codeUnits] is the number of code units in the original string.
+  static bool _isPlainAscii(List<int> bytes, int codeUnits) {
+    // Most non-ASCII code units will produce multiple bytes and make the text
+    // longer.
+    if (bytes.length != codeUnits) return false;
+
+    for (var byte in bytes) {
```

Comment:
totally optional but this could also be:
```dart
return bytes.every((b) => b & 0x80 == 0);
```

... instead of the loop. I guess it depends on how low level optimized this code is trying to be. (Fastest would be a `for (var i = 0, len = bytes.length; i < len; ++i)` style loop, IIRC)